[X86] Use shift+add/sub for vXi8 splat multiplies

grodranlorth opened the pull request:
Fixes #164200

~~I will create a separate PR to the `llvm-test-suite` repo for the microbenchmark for this change.~~ The benchmark is in https://github.com/llvm/llvm-test-suite/pull/316

In my experiments on an EC2 `c6i.4xl`, the change gives a small improvement for the `x86-64`, `x86-64-v2`, and `x86-64-v3` targets. It regresses performance on `x86-64-v4` (in particular, when the constant decomposes into two shifts). The performance summary follows:

```
$ ../MicroBenchmarks/libs/benchmark/tools/compare.py  benchmarks results-baseline-generic-v1.json results-opt-generic-v1.json  |tail -n1
OVERALL_GEOMEAN                         -0.2846         -0.2846             0             0             0             0
$ ../MicroBenchmarks/libs/benchmark/tools/compare.py  benchmarks results-baseline-generic-v2.json results-opt-generic-v2.json  |tail -n1
OVERALL_GEOMEAN                         -0.0907         -0.0907             0             0             0             0
$ ../MicroBenchmarks/libs/benchmark/tools/compare.py  benchmarks results-baseline-generic-v3.json results-opt-generic-v3.json  |tail -n1
OVERALL_GEOMEAN                         -0.1821         -0.1821             0             0             0             0
$ ../MicroBenchmarks/libs/benchmark/tools/compare.py  benchmarks results-baseline-generic-v4.json results-opt-generic-v4.json  |tail -n1
OVERALL_GEOMEAN                         +0.0190         +0.0190             0             0             0             0
```

grodranlorth committed:
[X86] Use shift+add/sub for vXi8 splat multiply

And avoid extending to 16 bits

grodranlorth committed:
[X86] lit tests for vXi8 shift+add/sub without width extension

arsenm commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -3413,6 +3419,59 @@ bool X86TargetLowering::decomposeMulByConstant(LLVMContext &Context, EVT VT,
 }
 
+TargetLowering::MulByConstInfo
+X86TargetLowering::getMulByConstInfo(EVT VT, const APInt &Constant) const {
```
I don't see why this is an overridable target hook; no target information is used here. This seems like it could just be separate utility function

grodranlorth replied:
Thank you, Matt. I used an overridable target hook as a way to make the optimization enabled only on x86, but I would be happy to convert it to a separate utility function and detect the target arch some other, simpler way, if possible.

arsenm replied:
>  I used an overridable target hook as a way to make the optimization enabled only on x86, 

This shouldn't be a goal. The project is worse for all of the hooks that only exist to avoid touching other targets

Commit:
fixup! [X86] Use shift+add/sub for vXi8 splat multiply

- Remove unnecessary target API
- Use existing decomposition logic

grodranlorth commented on llvm/test/CodeGen/X86/vector-mul-i8-decompose.ll:
```diff
@@ -0,0 +1,1231 @@
+; Test multiply by 6 = 4 + 2 = (1 << 2) + (1 << 1)
+; SSE2-LABEL: mul_v16i8_const6:
+; SSE2:       # %bb.0:
+; SSE2-NEXT:    movdqa %xmm0, %xmm1
```
Exact powers of 2 are no longer explicitly handled by this change, but already handled via existing code.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -4841,7 +4841,8 @@ template <class MatchContextClass> SDValue DAGCombiner::visitMUL(SDNode *N) {
     APInt MulC = ConstValue1.abs();
-    // The constant `2` should be treated as (2^0 + 1).
+    // Decompose 2^N ± 2^M as 2^(A+B) ± 2^B. The constant `2` should be
+    // treated as (2^0 + 1).
```
Necessary?

RKSimon commented on llvm/test/CodeGen/X86/vector-mul-i8-negative.ll:
```diff
@@ -0,0 +1,466 @@
+; To run all X86 CodeGen tests:
+;   llvm-lit llvm/test/CodeGen/X86/
+; To regenerate CHECK lines after modifying the test:
+;   python llvm/utils/update_llc_test_checks.py --llc-binary build/Release/bin/llc.exe llvm/test/CodeGen/X86/vector-mul-i8-negative.ll
```
remove all this

grodranlorth committed:
fixup! [X86] Use shift+add/sub for vXi8 splat multiply

- Second round of PR feedback

RKSimon reviewed (changes requested):
One last thing - please can you move the negative tests to the bottom of the vector-mul-i8-decompose.ll - no need to split them

grodranlorth committed:
Move vXi8 tests into a single file

grodranlorth committed:
fixup! [X86] Use shift+add/sub for vXi8 splat multiply

- remove redundant comment

RKSimon commented on llvm/test/CodeGen/X86/vector-mul-i8-decompose.ll:
```diff
@@ -0,0 +1,1808 @@
+;; v16i8 Tests (128-bit vectors) - Difference of two powers of 2
+;; ============================================================================
+
+; Test multiply by 6 = 8 - 2 = (1 << 3) - (1 << 1)
```
None of these seems to be use (v)psubb?
```suggestion
; Test multiply by 6 = 8 - 2 = (1 << 3) - (1 << 1) = (1 << 2) + (1 << 1)
```

grodranlorth committed:
RKSimon's test feedback

- Fixed tests to actually exercise the negative branch
- Use "splat"
- Fix test comments

RKSimon reviewed (approved):
LGTM - cheers

RKSimon committed:
Merge branch 'main' into shiftopt
